probe-rs/src/session.rs

```diff
@@ -432,6 +423,111 @@ impl Session {
         Ok(interface)
     }
 
+    fn reattach_arm_interface(
+        interface: &mut Box<dyn ArmProbeInterface>,
+        debug_sequence: &Arc<dyn ArmDebugSequence>,
+    ) -> Result<(), Error> {
+        use crate::DebugProbe;
+
+        // Temporarily swap out the interface for a Dummy
+        let tmp_interface = Box::new(FakeProbe::default())
+            .try_get_arm_interface()
+            .unwrap();
+        let mut tmp_interface = tmp_interface
+            .initialize(DefaultArmSequence::create())
+            .unwrap();
+
+        std::mem::swap(interface, &mut tmp_interface);
+
+        log::debug!("Re-attaching Probe");
+        let mut probe = tmp_interface.close();
+        probe.detach()?;
+        probe.attach_to_unspecified()?;
+
+        let new_interface = probe.try_into_arm_interface().map_err(|(_, err)| err)?;
+
+        tmp_interface = new_interface.initialize(debug_sequence.clone())?;
+        // swap it back
+        std::mem::swap(interface, &mut tmp_interface);
+
+        log::debug!("Probe re-attached");
+        Ok(())
+    }
+
+    /// Check if the connected device has a custom erase sequence defined
+    pub fn has_custom_erase_all(&self) -> bool {
+        match &self.target.debug_sequence {
+            DebugSequence::Arm(seq) => seq.custom_erase_sequence().is_some(),
+            DebugSequence::Riscv(_) => {
+                log::info!("Custom Erase Sequences are not implmented for RISCV");
+                false
+            }
+        }
+    }
+
+    /// Erase all flash memory using the Device's Custom Erase Sequence if any
+    ///
+    /// # Returns
+    /// Ok(()) if the device provides a custom erase sequence and it succeeded.
+    ///
+    /// # Errors
+    /// NotImplemented if no custom erase sequence exists
+    /// Err(e) if the custom erase sequence failed
+    pub fn custom_erase_all(&mut self, permissions: &Permissions) -> Result<(), Error> {
+        match &mut self.interface {
+            ArchitectureInterface::Arm(interface) => {
+                let debug_sequence = match &self.target.debug_sequence {
+                    DebugSequence::Arm(seq) => seq.clone(),
+                    DebugSequence::Riscv(_) => {
+                        unreachable!("This should never happen. Please file a bug if it does.")
+                    }
+                };
+
+                if let Some(custom_erase) = debug_sequence.custom_erase_sequence() {
+                    let config = self.target.cores[0].clone();
```

Comment:
Not sure if the core shouldn't be an argument to the function or if erase will always work on core 0.

Reply:
I'm not sure either and need your expertise here.  I would guess that a chip erase command would always be at least available on Core 0, but perhaps available on all cores. I also assume there is always just one flash array that is erased (or do you know of any chips with different flash arrays per core?)

Reply:
I would say **erase** is always available on core 0.
The question is if we consider special unlock erase sequences here. For nRFs there is a special unlock erase that you can perform through an AP, so core access is not even the correct thing.
I am also okay with breaking API between releases. So I think we should just keep this for now and adapt if we run into a different need instead of contemplating forever :)

Reply:
I think we can change the API to only pass the arm interface instead.  The Erase implementation can then convert it to the appropriate Memory interface if needed or just use the AP directly. Then we don't have to do any guessing here :-)